lcli/src/etl/block_efficiency.rs

```diff
@@ -0,0 +1,317 @@
+    let offline_window: u64 = matches
+        .value_of("offline_window")
+        .unwrap_or("3")
+        .parse()
+        .map_err(|e| format!("{:?}", e))?;
+
+    if start_epoch == 0 {
+        return Err("start_epoch cannot be 0.".to_string());
+    }
+    let initialization_epoch: Epoch = start_epoch - 1;
+    let end_epoch: Epoch = clap_utils::parse_required(matches, "end-epoch")?;
+
+    if end_epoch < start_epoch {
+        return Err("start_epoch must be smaller than end_epoch".to_string());
+    }
+
+    let relevant_epochs: Vec<Epoch> = (initialization_epoch.as_u64()..=end_epoch.as_u64())
+        .map(Epoch::new)
+        .collect();
+
+    let mut available_attestations_set: HashSet<UniqueAttestation> = HashSet::new();
+    let mut included_attestations_set: HashMap<UniqueAttestation, InclusionDistance> =
+        HashMap::new();
+
+    // Build validator set HashMap<Index, EpochOfMostRecentAttestation)>
+    // This allows a 'best effort' attempt to normalize block efficiencies.
+    let mut online_validator_set: HashMap<ValidatorIndex, Epoch> = HashMap::new();
+
+    let mut proposer_map: HashMap<Slot, ProposerInfo> = HashMap::new();
+
+    let mut file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(output_path)
+        .map_err(|e| format!("Unable to open file: {}", e))?;
+
+    write!(file, "slot,proposer,available,included,offline,graffiti").unwrap();
+
+    // Initialize API.
+    let endpoint = matches
+        .value_of("endpoint")
+        .unwrap_or("http://localhost:5052/");
+    let node = BeaconNodeHttpClient::new(
+        SensitiveUrl::parse(endpoint).map_err(|_| "Unable to parse endpoint.".to_string())?,
+        Timeouts::set_all(SECONDS_PER_SLOT),
+    );
+
+    // Loop over epochs.
+    for epoch in relevant_epochs {
+        if epoch != initialization_epoch {
+            eprintln!("Analysing epoch {}...", epoch);
+        } else {
+            eprintln!("Initializing...");
+        }
+        let mut epoch_data: Vec<(Slot, Option<ProposerInfo>, usize, usize)> = Vec::new();
+
+        // Current epochs available attestations set
+        let committee_data = match get_epoch_committee_data::<T>(&node, epoch).await {
+            Ok((committee_data, committee_info)) => (committee_data, committee_info),
+            Err(e) => return Err(e),
+        };
+
+        // Ensure available attestations don't exceed the possible amount of attestations
+        // as determined by the committee size/number.
+        // This is unlikely to happen, but not impossible.
+        let max_possible_attesations =
+            committee_data.1.validators_per_committee * committee_data.1.number_of_committees;
+
+        // Get number of active validators.
+        let active_validators =
+            get_validator_set_len::<T>(&node, epoch.start_slot(T::slots_per_epoch())).await?;
+
+        let start_slot = epoch.start_slot(T::slots_per_epoch());
+        for step in 0..32 {
```

Comment:
```suggestion
        for slot in epoch.slot_iter(T::slots_per_epoch()) {
```

Doesn't make any difference at runtime, but it uses a snazzy iter I made :sweat_smile:

Reply:
Ah nice. I did know about this, no sure why I didn't use it :sweat_smile: